### ISSUE TITLE
Best practice to wrap YOLOv8 with FastAPI

### ISSUE THREAD

**Issue by darouwan** [ultralytics/ultralytics#8235]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have read the document here [https://docs.ultralytics.com/guides/yolo-thread-safe-inference/?h=thread](url), and it suggest to initialize the YOLO instance inside of thread to ensure thread safety. 

Now I am working on building a model web service with FastAPI.  So what is the best practice to initial the YOLO instance in FastAPI? If my container only have 1 GPU, can I share 1 GPU among multiple YOLO instance with multiple FastAPI workers?

**Comment by glenn-jocher** [comment]
Hello! Wrapping YOLOv8 with FastAPI for a model web service is a great idea. Here's a simple way to do it:

1. Initialize the YOLO model within your FastAPI endpoint function. This ensures that each request gets a fresh model instance, which is thread-safe.
2. If you have a single GPU, you can indeed share it among multiple YOLO instances. However, be mindful of the GPU memory usage, as each instance will consume memory. You might need to manage the number of concurrent requests to avoid memory overflow.

Here's a small code snippet to get you started:

```python
from fastapi import FastAPI
from ultralytics import YOLO

app = FastAPI()
 @darouwan.post("/predict")
def predict(image: bytes):
    model = YOLO('yolov8n.pt', device='0')  # Load YOLO model on GPU 0
    results = model.predict(image)
    return results.pandas().xyxy[0].to_json(orient="records")
```

This example assumes you're sending image data directly to the endpoint. Adjust the `predict` function parameters as needed for your use case. Good luck with your FastAPI service! 🚀

**Comment by darouwan** [comment]
Thanks for your quick resonse @glenn-jocher 
If I initialize the yolo model every time when receive new request, will it be to slow for every prediction request?

**Comment by glenn-jocher** [comment]
Hi there! Initializing the YOLO model for each request can indeed add some overhead, but it's generally quite fast and ensures thread safety. If you're concerned about speed, you could consider loading the model once at startup and then using it to predict in a thread-safe manner. Just make sure to manage concurrent access properly. Here's a quick example:

```python
from fastapi import FastAPI
from ultralytics import YOLO

app = FastAPI()
model = YOLO('yolov8n.pt', device='0')  # Load model at startup
 @darouwan.post("/predict")
def predict(image: bytes):
    results = model.predict(image)
    return results.pandas().xyxy[0].to_json(orient="records")
```

This way, you load the model only once, and the prediction requests should be faster. Happy coding! 😊🚀

**Comment by AlexPasqua** [comment]
Hello @darouwan and @glenn-jocher ,
am I wrong or the `results.pandas().xyxy[0]` was a feature of YOLOv5 which is not a=currently available for YOLOv8?

In particular, in my case results is a list (which obviously doesn't have the attribute `pandas()`), but even accessing the elements of such list, the situation doesn't change.
For reference, this is what I do:
```python
import cv2
import numpy as np
from ultralytics import YOLO
from fastapi import FastAPI, File, UploadFile


@app.post("/detect")
async def detect(file: UploadFile):
    # Process the uploaded image for object detection
    image_bytes = await file.read()
    image = np.frombuffer(image_bytes, dtype=np.uint8)
    image = cv2.imdecode(image, cv2.IMREAD_COLOR)

    # Perform object detection with YOLOv8
    detections = model.predict(image)

    return {'detections': detections.pandas().xyxy[0].to_json()}
```

The only way to print the results so far is substituting the above `return` statement with
```python
return {'detections': detections[0].tojson()}
```
However, the formatting is not great (e.g., it explicitly shows all the various `\n` and so on).

**Comment by glenn-jocher** [comment]
Hello @AlexPasqua,

Thank you for reaching out! You're correct that the `results.pandas().xyxy[0]` method was a feature in YOLOv5 and is not directly available in YOLOv8. In YOLOv8, the results are returned as a list of `Results` objects, and you can access the bounding box data differently.

Here's an updated version of your code that should work with YOLOv8:

```python
import cv2
import numpy as np
from ultralytics import YOLO
from fastapi import FastAPI, File, UploadFile

app = FastAPI()
model = YOLO('yolov8n.pt', device='0')  # Load model at startup

@app.post("/detect")
async def detect(file: UploadFile):
    # Process the uploaded image for object detection
    image_bytes = await file.read()
    image = np.frombuffer(image_bytes, dtype=np.uint8)
    image = cv2.imdecode(image, cv2.IMREAD_COLOR)

    # Perform object detection with YOLOv8
    detections = model.predict(image)

    # Extract bounding box data
    boxes = detections[0].boxes.xyxy.cpu().numpy()
    scores = detections[0].boxes.conf.cpu().numpy()
    classes = detections[0].boxes.cls.cpu().numpy()

    # Format the results as a list of dictionaries
    results = []
    for box, score, cls in zip(boxes, scores, classes):
        results.append({
            'x1': box[0],
            'y1': box[1],
            'x2': box[2],
            'y2': box[3],
            'confidence': score,
            'class': int(cls)
        })

    return {'detections': results}
```

This code extracts the bounding box coordinates, confidence scores, and class indices from the `Results` object and formats them into a list of dictionaries, which should be easier to work with and more readable.

If you encounter any issues, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, feel free to provide a minimum reproducible example so we can investigate further. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

I hope this helps! Let us know if you have any further questions. 😊

**Comment by AlexPasqua** [comment]
Hi @glenn-jocher , thanks for the quick response! 😄 
The only minor change I had to make to your code in order to make it work for me was to add a `.astype(float32)` to every float I was saving in results, so something like:
```python
results = []
for box, score, cls in zip(boxes, scores, classes):
    results.append({
        'x1': box[0].astype(float32),
        'y1': box[1].astype(float32),
        'x2': box[2].astype(float32),
        'y2': box[3].astype(float32),
        'confidence': score.astype(float32),
        'class': int(cls)
    })
```
It looked like the json encoder wasn't too happy about numpy types..

Cheers! 🚀

**Comment by glenn-jocher** [comment]
Hi @AlexPasqua,

Thank you for your feedback and for sharing the solution that worked for you! 😊

It's great to hear that you were able to resolve the issue by converting the numpy types to standard Python types. JSON encoding can indeed be particular about data types, and your approach to use `.astype(float)` ensures compatibility.

Here's the updated snippet for clarity:

```python
results = []
for box, score, cls in zip(boxes, scores, classes):
    results.append({
        'x1': float(box[0]),
        'y1': float(box[1]),
        'x2': float(box[2]),
        'y2': float(box[3]),
        'confidence': float(score),
        'class': int(cls)
    })
```

This should help anyone else facing similar issues. If you encounter any more questions or need further assistance, feel free to reach out. Happy coding! 🚀

**Comment by AlexPasqua** [comment]
> You're correct that the `results.pandas().xyxy[0]` method was a feature in YOLOv5 and is not directly available in YOLOv8.

At this point, wouldn't it be nice to have this feature is YOLOv8 too?
I think extracting the boxes' info in this way is fairly common, and atm it requires quite a bit of "manual" work and lines of code. I think it'd be good to have a more automatic way of getting such info.

If you think so, I could open an issue (feature request) to discuss this, and eventually a PR to implement it! 😄

**Comment by glenn-jocher** [comment]
Hello,

Thank you for your insightful suggestion! 😊

You're absolutely right that having a `results.pandas().xyxy[0]` method in YOLOv8 would streamline the process of extracting bounding box information, making it more user-friendly and efficient. This feature was indeed quite handy in YOLOv5, and bringing it to YOLOv8 could benefit many users.

We encourage you to open a feature request on our GitHub repository to discuss this further. Your input and potential contribution through a PR would be greatly appreciated by the community. To get started, you can create a new issue [here](https://github.com/ultralytics/ultralytics/issues) and outline the feature you'd like to see implemented.

Thank you for your enthusiasm and willingness to contribute! 🚀
